File: E-mart_java/Backend/src/main/java/com/example/Controllers/SubcategoryController.java
package com.example.Controllers;


import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import com.example.Models.Subcategory;
import com.example.Services.SubcategoryService;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/subcategories")
@CrossOrigin(origins = "http://localhost:5173")
public class SubcategoryController {

    @Autowired
    private SubcategoryService subcategoryService;

    @GetMapping
    public List<Subcategory> getAllSubcategories() {
        return subcategoryService.getAllSubcategories();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Subcategory> getSubcategoryById(@PathVariable int id) {
        Optional<Subcategory> subcategory = subcategoryService.getSubcategoryById(id);
        return subcategory.map(ResponseEntity::ok).orElseGet(() -> ResponseEntity.notFound().build());
    }

    @GetMapping("/category/{categoryId}")
    public List<Subcategory> getSubcategoriesByCategory(@PathVariable int categoryId) {
        return subcategoryService.getSubcategoriesByCategory(categoryId);
    }

}

File: E-mart_.Net/Backend/Controllers/SubCategoryController.cs
using E_mart.Services;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using System.Threading.Tasks;

namespace E_mart.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class SubCategoryController : ControllerBase
    {
        private readonly ISubCategoryService _subCategoryService;

        public SubCategoryController(ISubCategoryService subcategoryService)
        {
            _subCategoryService = subcategoryService;
        }

        [HttpGet]
        public async Task<IActionResult> GetAllSubcategories()
        {
            var subcategories = await _subCategoryService.GetAllSubcategories(); 
            if (subcategories == null || !subcategories.Any()) 
            {
                return NotFound();
            }
            return Ok(subcategories);
        }

        [HttpGet("{subcategoryid}")]
        public async Task<IActionResult> GetSubcategoryById(int subcategoryid)
        {
            var subcategory = await _subCategoryService.GetSubcategoryById(subcategoryid); 
            if (subcategory == null)
            {
                return NotFound();
            }
            return Ok(subcategory);
        }

        [HttpGet("category/{categoryid}")]
        public async Task<IActionResult> GetSubcategoriesByCategory(int categoryid)
        {
            var subcategories = await _subCategoryService.GetSubcategoriesByCategory(categoryid); 
            if (subcategories == null || !subcategories.Any()) 
            {
                return NotFound();
            }
            return Ok(subcategories);
        }
    }
}
